docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,319 @@
+cached table descriptors.
+
+# Motivation
+
+Table descriptors contain the schema for a single table and are
+utilized by nearly every SQL operation. Fast access to the table
+descriptor is critical for good performance. Reading the table
+descriptor from the KV store on every operation adds significant
+latency.
+
+Table descriptors are currently distributed to every node in the
+cluster via gossipping of the system config (see
+[schema_distribution](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/schema_gossip.md)). Unfortunately,
+it is not safe to use these gossipped table descriptors in almost any
+circumstance. Consider the statements:
+
+```sql
+  CREATE TABLE test (key INT PRIMARY KEY, value INT);
+  CREATE INDEX foo ON test (value);
+  INSERT INTO test VALUES (1, 2);
+```
+
+Depending on when the gossip of the schema change is received, the
+`INSERT` statement may either see no cached descriptor for the `test`
+table (which is safe because it will fall back to reading from the KV
+store), the descriptor as it looks after the `CREATE TABLE` statement
+or the descriptor as it looks after the `CREATE INDEX` statement. If
+the `INSERT` statement does not see the `CREATE INDEX` statement then
+it will merrily insert the values into `test` without creating the
+entry for the `foo` index, thus leaving the table data in an
+inconsistent state.
+
+It is easy to show similar problematic sequences of operations for
+`DELETE`, `UPDATE` and `SELECT`. Fortunately, a solution exists in the
+literature: break up the schema change into discrete steps for which
+any two consecutive versions of the schema are safe for concurrent
+use. For example, to add an index to a table we would perform the
+following steps:
+
+* Add the index to the table descriptor but mark it as delete-only:
+  update and delete operations which would delete entries for the
+  index do so, but insert and update operations which would add new
+  elements and read operations do not use the index.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster and all uses of the previous version to finish.
+* Mark the index as write-only: insert, update and delete operations
+  would add entries for the index, but read operations would not use
+  the index.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster and all uses of the previous version to finish.
+* Backfill the index.
+* Mark the index in the table descriptor as read-write.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster and all uses of the previous version to finish.
+
+This RFC is focused on how to wait for the table descriptor change
+propagate to all nodes in the cluster. Additionally, we need
+mechanisms to ensure that multiple updates to a table descriptor are
+not performed concurrently and that write operations never commit once
+a table descriptor is too old (i.e. not one of the two most recent
```

Comment:
Can you remove, "write operations never commit once a table descriptor is too old" 
and change the first line to.

This RFC is focused on how to wait for the table descriptor change to
+propagate to all nodes in the cluster and all the uses of the previous version to finish.